keps/sig-cluster-lifecycle/addons/20200121-manifest-bundle.md

```diff
@@ -0,0 +1,189 @@
+
+By standardizing on a container format for this data, we get many other features for free, such as: identity, distribution, replication, deduplication, signing, and cluster ingress.
+
+## Goals
+
+* Define a convention for storing manifests and metadata with a container image
+* Build and push metadata using standard container tooling (docker cli, podman, etc) to common container image registries (Artifactory, docker hub, quay, google's registry, amazon's registry)
+* No union filesystem should be required to consume metadata
+* Have a simple mechanism to apply a bundle to a kubernetes cluster
+
+## Non-Goals
+
+* Require OCI registries that support any non-standard media types (i.e. anything other than docker and layer image oci)
+* Build on-cluster tooling to interact with bundles
+
+## Proposal
+
+We would like to make a distinction between:
+
+* **Manifests** * data which can be applied to a cluster, either with or without pre-processing. Examples: json or yaml manifests, helm charts, kustomize bundles.
+* **Metadata** * information which will not be directly applied to a cluster, but which will otherwise affect the processing or application of a bundle. 
+
+## User Stories
+
+### Build, Push, Pull Manifest Bundle
+
+As a manifest author, I would like to store manifests and metadata in a format compatible with container registries.
+
+Constraints:
+
+* An operator bundle (including both manifests and metadata) should be identifiable using a single versioned identifier.
+
+## Implementation Details/Notes/Constraints
+
+The initial implementation target will be Docker v2-2 manifests, manifest-lists, and client support, for maximum compatibility with existing tooling. OCI is a future target, but avoided for now due to lack of tooling support. 
+
+Labels are used to identify the contents. The approach is similar to OCI mediatypes, but is supported by existing tooling.
+
+The following label convention is used to annotate the bundle image:
+
+* `<namespace>.mediatype.<version>=<identifier>` is used to identify the top-level format of the bundle. For example, this may indicate that the bundle contains a kustomization file and kustomize manifests.
+* `<namespace>.bundle.manifests.<version>=<path>` reflects the path in the image to a directory that contains manifests. 
+* `<namespace>.bundle.metadata.<version>=<path>` reflects the path in the image to a directory that contains metadata.
+* Any additional `<namespace>.bundle.<identifier>=<value>` may be used to indicate additional properties of the manifest bundle. It may be useful to denormalize information that would otherwise be stored in the metadata directory, so that tooling can read it without unpacking a full image.
+These labels should also be replicated in a well-known location within the image, metadata/annotations.yaml:
+
+```yaml
+annotations:
+  <namespace>.mediatype.<version>: <identifier>
+  <namespace>.bundle.manifests.<version>: <path>
```

Comment:
Q: If it's mirrored into metadata/annotations.yaml, is it important to expose the paths in the image labels?  I imagine if I care about the path, I'm fetching the image anyway?  (And I don't think I can partially fetch an image?)

Reply:
The idea (and hope) was that eventually we could have something on-cluster that can read image labels directly. Right now the labels are useful for off-cluster tooling that can read the labels, though that tooling would need to assume that the labels are being kept up-to-date.

One idea would be something like the image volume driver providing an option to have the mount location specified by an image label. 

```yaml
  volumes:
  - name: data
    csi:
      driver: image.csi.k8s.io
      volumeAttributes:
          label: <namespace>.bundle.manifests.<version>
          image: myimage:tag
```